### ISSUE TITLE
dynamic functions converts the whole route to dynamic in PPR

### ISSUE THREAD

**Issue by A7med3bdulBaset** [vercel/next.js#59429]
### Link to the code that reproduces this issue

https://github.com/A7med3bdulBaset/ppr-reproduce

### To Reproduce

1. `next build`
2. http://localhost:3000/ppr

### Current vs. Expected behavior

I was trying partial pre-rendering, made a page with two RSC within `<Suspense>` one of the with fallback. All of `layout.tsx`, `page.tsx`, RSCs, and Suspense fallback render `new Date().toLocaleTimeString()` and `Math.random`. e.g.
```tsx
const time = new Date().toLocaleTimeString("en-US", {
  timeZone: "America/New_York",
});

return (
<div>
  <p>Time: {time}</p>
  <p>Random: {(Math.random() * 10).toFixed(1)}</p>
</div>
);
```

When I build the app I can see that it's static as expected:

```md
build: Route (app)                              Size     First Load JS
build: ┌ ○ /                                    6.93 kB        90.4 kB
build: ├ ○ /_not-found                          886 B          84.3 kB
build: └ ○ /ppr                                 136 B          83.6 kB
build:  // .........................
build: ○  (Static)  prerendered as static content
```

Let's start trying PPR by calling `unstable_noStore` in a component:
```tsx
// import { headers } from "next/headers"
import { unstable_noStore } from "next/cache";

export async function AsyncComponent() {
  unstable_noStore();
  // headers()

  const time = new Date().toLocaleTimeString("en-US", {
    timeZone: "America/New_York",
  });

  return (
    <div className="bg-zinc-100 border border-zinc-300 transition hover:border-zinc-500 p-6">
      <h2>Dynamic Component</h2>
      <p>Time: {time}</p>
      <p>Random: {(Math.random() * 10).toFixed(1)}</p>
    </div>
  );
}
```
When building the app I get it ppr as expected
```md
build: Route (app)                              Size     First Load JS
build: ┌ ○ /                                    6.93 kB        90.4 kB
build: ├ ○ /_not-found                          886 B          84.3 kB
build: └ ◐ /ppr                                 136 B          83.6 kB
build: // ................................
build: ○  (Static)             prerendered as static content
build: ◐  (Partial Prerender)  prerendered as static HTML with dynamic server-streamed content
```

That's awesome!

But when running `next start`, The content dynamic suspense is no longer static. `unstable_noStore()` as well as `headers()` make the whole page dynamic, whenever I refresh the page, It's dynamically changes.

I expect to see `new Date().toLocaleTimeString()` calls outside dynamic components to be the build time not the refresh page time. The only content should be updated is the component that calls `noStore`.

I also tested the same steps at https://github.com/vercel-labs/next-partial-prerendering It's the same.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Home Single Language
Binaries:
  Node: 20.0.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.0.5-canary.4
  eslint-config-next: 14.0.3
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

**Comment by A7med3bdulBaset** [comment]
Thanks @wyattjoh for your reply

> the user receives some of the static part of the page as soon as possible so that the browser can start downloading assets on the page and can avoid the "white screen of death" as much as possible.

As far as I know, you're describing Streaming here. How PPR is different from Streaming?

> So in this context, the point of Partial Pre Rendering (PPR) is not to create a static part of the shell that never re-renders.

What I understand about PPR, it generates the static parts of the page as well as the Suspense fallback (at the build time) and dynamically build the dynamic parts (at the request time) not dynamically build the whole page (like without PPR). 

I'm quoting from [Sebmarkbage and Malte's blog about PPR](https://vercel.com/blog/partial-prerendering-with-next-js-creating-a-new-default-rendering-model):

> With PPR enabled, this page generates a static shell based on your <Suspense /> boundaries. The fallback provided to React Suspense is prerendered.
>
> Suspense fallbacks in the shell are then replaced with dynamic components, like reading cookies to determine the cart, or showing a banner based on the user.

And [partialprerendering.com](https://www.partialprerendering.com/):

> ## How it works:
> 
> - A static route shell is served immediately, this makes the initial load fast.
> - The shell leaves holes where dynamic content (that might be slower) will be streamed in to minimize the perceived overall page load time.

My understanding is when using `Date.now()` in the `layout.tsx`, outside `<Suspense />` or at Suspense fallback, I expect to have the same value whenever I refresh the page (build time in ms).

---

> How PPR currently works is that an initial pass during page generation is done where we mark the render as being a target for "static generation", and as a result, will cause dynamic functions (like unstable_noStore()) to throw a special error for us that's caught by suspense that indicates that this particular component has postponed due to dynamic access, and as a result, should not have its contents cached. When this page is later resumed by navigating to the page, we actually re-render the entire page, not just from the points where the static rendering postponed. This actually allows for neat situations where in your case, the resume is actually patching the running page with the fresher content. This is mostly a result of methods like Math.random() and Date.now() not invoking our dynamic error as fetch, headers, or unstable_noStore() does.

If that is the case, How it's different from Dynamic Rendering? They both build the whole route at the request time when using dynamic functions such as `unstable_noStore()`. What is the difference and the benefit of PPR over Dynamic Rendering.

---

> TLDR; if this isn't the behavior you're looking for, place a call to unstable_noStore() around calls to Math.random() or Date.now() to ensure they'll only show the dynamic content on resume.

You might missed my point. What I need/expect is having the fresh dynamic content only within the component that uses dynamic function. Please reach out to the reproduction or look at this example

```tsx
// app/page.tsx

export default function Page() {
  return (
    <div>
      <p>This should be static {Date.now()}</p>
      <Suspense fallback={<StaticFallback />}>
        <DynamicContent />
      </Suspense>
    </div>
  )
}
```
```tsx
// app/dynamicContent.tsx

export default function DynamicContent() {
  cookies()
  return <p>This should be dynamic {Date.now()}</p>
}
```

I expect the `Date.now()` in the second <p /> won't be the same as (or neer to the first). The first date should be the build time, and the second should be the request time. When I refresh the page. I expect the first paragraph to not have different value. If it changes so the page is Dyanamically Rendered not Partially Prerendered.
